File: src/drod/stat.rs
use std::fmt::{Display, Formatter, Result};
use std::ops::{AddAssign, Neg, Sub};

bitflags! {
    #[derive(Default)]
    pub(super) struct PlayerBehavior: u32 {
        const HAS_WEAPON                = 0b000001;
        const DOUBLE_GR_WEAPON          = 0b000010;
        const DOUBLE_GR_ACCESSORY       = 0b000100;
        const DOUBLE_REP_ACCESSORY      = 0b001000;
        const DOUBLE_ATK_AGAINST_GOBLIN = 0b010000;
        const DOUBLE_ATK_AGAINST_WYRM   = 0b100000;
        const WEAPON_ATTR = Self::HAS_WEAPON.bits | Self::DOUBLE_GR_WEAPON.bits | Self::DOUBLE_ATK_AGAINST_GOBLIN.bits | Self::DOUBLE_ATK_AGAINST_WYRM.bits;
        const ACCESSORY_ATTR = Self::DOUBLE_GR_ACCESSORY.bits | Self::DOUBLE_REP_ACCESSORY.bits;
    }
}

impl Display for PlayerBehavior {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        let mut behavior_str = "Behavior:".to_string();
        if self.contains(PlayerBehavior::HAS_WEAPON) {
            behavior_str += " HAS_WEAPON";
        }
        if self.contains(PlayerBehavior::DOUBLE_GR_WEAPON) {
            behavior_str += " DOUBLE_GR_WEAPON";
        }
        if self.contains(PlayerBehavior::DOUBLE_GR_ACCESSORY) {
            behavior_str += " DOUBLE_GR_ACCESSORY";
        }
        if self.contains(PlayerBehavior::DOUBLE_REP_ACCESSORY) {
            behavior_str += " DOUBLE_REP_ACCESSORY";
        }
        if self.contains(PlayerBehavior::DOUBLE_ATK_AGAINST_GOBLIN) {
            behavior_str += " DOUBLE_ATK_AGAINST_GOBLIN";
        }
        if self.contains(PlayerBehavior::DOUBLE_ATK_AGAINST_WYRM) {
            behavior_str += " DOUBLE_ATK_AGAINST_WYRM";
        }

        write!(f, "{}", behavior_str)
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub(super) struct StatDiff {
    pub(super) behavior: PlayerBehavior,
    pub(super) hp: i32,
    pub(super) atk: i32,
    pub(super) def: i32,
    pub(super) equip_atk: i32,
    pub(super) equip_def: i32,
    pub(super) gr: i32,
    pub(super) rep: i32,
    yk: i32,
    gk: i32,
    bk: i32,
    sk: i32,
}

impl StatDiff {
    pub(super) fn is_free(&self) -> bool {
        self.hp >= 0
            && self.atk >= 0
            && self.def >= 0
            && self.equip_atk >= 0
            && self.equip_def >= 0
            && self.gr >= 0
            && self.rep >= 0
            && self.yk >= 0
            && self.gk >= 0
            && self.bk >= 0
            && self.sk >= 0
    }
}

impl AddAssign for StatDiff {
    fn add_assign(&mut self, other: Self) {
        self.behavior ^= other.behavior;
        self.hp += other.hp;
        self.atk += other.atk;
        self.def += other.def;
        self.equip_atk += other.equip_atk;
        self.equip_def += other.equip_def;
        self.gr += other.gr;
        self.rep += other.rep;
        self.yk += other.yk;
        self.gk += other.gk;
        self.bk += other.bk;
        self.sk += other.sk;
    }
}

impl Neg for StatDiff {
    type Output = Self;

    fn neg(self) -> Self {
        assert!(self.behavior.is_empty());
        Self {
            behavior: self.behavior,
            hp: -self.hp,
            atk: -self.atk,
            def: -self.def,
            equip_atk: -self.equip_atk,
            equip_def: -self.equip_def,
            gr: -self.gr,
            rep: -self.rep,
            yk: -self.yk,
            gk: -self.gk,
            bk: -self.bk,
            sk: -self.sk,
        }
    }
}

// HP is shifted by 1 so that 0 is considered alive.
// This change makes code cleaner.
#[derive(Clone, Copy, Debug, Default)]
pub struct Player {
    behavior: PlayerBehavior,
    pub(super) hp: i32,
    atk: i32,
    def: i32,
    equip_atk: i32,
    equip_def: i32,
    gr: i32,
    rep: i32,
    yk: i32,
    gk: i32,
    bk: i32,
    sk: i32,
}

impl Player {
    // Calculates score scaled by 1000
    pub(super) fn score(&self) -> i32 {
        // Default score formula for ToTS floors 25 and 49
        // (self.hp + 1) * 25
        //     + (self.atk * 5 + self.yk * 10 + self.gk * 20 + self.bk * 30 + self.sk * 30) * 1000
        //     + self.def * 1000 / 10 * 3;

        // Default score formula for DROD RPG
        (self.hp + 1) * 25
            + (self.atk * 5
                + self.def * 3
                + self.yk * 10
                + self.gk * 20
                + self.bk * 30
                + self.sk * 30)
                * 1000
    }

    fn join(&mut self, other: Self) {
        self.behavior |= other.behavior;
        self.hp = self.hp.max(other.hp);
        self.atk = self.atk.max(other.atk);
        self.def = self.def.max(other.def);
        self.equip_atk = self.equip_atk.max(other.equip_atk);
        self.equip_def = self.equip_def.max(other.equip_def);
        self.gr = self.gr.max(other.gr);
        self.rep = self.rep.max(other.rep);
        self.yk = self.yk.max(other.yk);
        self.gk = self.gk.max(other.gk);
        self.bk = self.bk.max(other.bk);
        self.sk = self.sk.max(other.sk);
    }

    pub(super) fn dominate(&self, other: &Self) -> bool {
        self.behavior.contains(other.behavior)
            && self.hp >= other.hp
            && self.atk >= other.atk
            && self.def >= other.def
            && self.equip_atk >= other.equip_atk
            && self.equip_def >= other.equip_def
            && self.gr >= other.gr
            && self.rep >= other.rep
            && self.yk >= other.yk
            && self.gk >= other.gk
            && self.bk >= other.bk
            && self.sk >= other.sk
    }
}

impl From<StatDiff> for Player {
    fn from(stat: StatDiff) -> Self {
        Self {
            behavior: stat.behavior,
            hp: stat.hp,
            atk: stat.atk,
            def: stat.def,
            equip_atk: stat.equip_atk,
            equip_def: stat.equip_def,
            gr: stat.gr,
            rep: stat.rep,
            yk: stat.yk,
            gk: stat.gk,
            bk: stat.bk,
            sk: stat.sk,
        }
    }
}

impl From<CombatStat> for Player {
    fn from(stat: CombatStat) -> Self {
        Self {
            behavior: stat.behavior,
            atk: stat.atk,
            def: stat.def,
            equip_atk: stat.equip_atk,
            equip_def: stat.equip_def,
            ..Self::default()
        }
    }
}

impl AddAssign<StatDiff> for Player {
    fn add_assign(&mut self, other: StatDiff) {
        self.behavior &= other.behavior;
        self.hp += other.hp;
        self.atk += other.atk;
        self.def += other.def;
        self.equip_atk += other.equip_atk;
        self.equip_def += other.equip_def;
        self.gr += other.gr;
        self.rep += other.rep;
        self.yk += other.yk;
        self.gk += other.gk;
        self.bk += other.bk;
        self.sk += other.sk;
    }
}

impl Sub<StatDiff> for Player {
    type Output = Self;

    fn sub(self, other: StatDiff) -> Self {
        Self {
            behavior: self.behavior ^ other.behavior,
            hp: self.hp - other.hp,
            atk: self.atk - other.atk,
            def: self.def - other.def,
            equip_atk: self.equip_atk - other.equip_atk,
            equip_def: self.equip_def - other.equip_def,
            gr: self.gr - other.gr,
            rep: self.rep - other.rep,
            yk: self.yk - other.yk,
            gk: self.gk - other.gk,
            bk: self.bk - other.bk,
            sk: self.sk - other.sk,
        }
    }
}

impl Display for Player {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        write!(
            f,
            "{{ {}, HP: {}, ATK: {}, DEF: {}, GR: {}, REP: {}, YK: {}, GK: {}, BK: {}, SK: {}}}",
            self.behavior,
            self.hp + 1,
            self.atk,
            self.def,
            self.gr,
            self.rep,
            self.yk,
            self.gk,
            self.bk,
            self.sk
        )
    }
}

#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq)]
pub(super) struct CombatStat {
    pub(super) behavior: PlayerBehavior,
    pub(super) atk: i32,
    pub(super) def: i32,
    pub(super) equip_atk: i32,
    pub(super) equip_def: i32,
}

impl From<Player> for CombatStat {
    fn from(stat: Player) -> Self {
        Self {
            behavior: stat.behavior,
            atk: stat.atk,
            def: stat.def,
            equip_atk: stat.equip_atk,
            equip_def: stat.equip_def,
        }
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub(super) struct ProbeStat {
    pub(super) diff: StatDiff,
    pub(super) req: Player,
    pub(super) damage: i32,
}

impl AddAssign for ProbeStat {
    fn add_assign(&mut self, other: Self) {
        self.req.join(other.req - self.diff);
        self.diff += other.diff;
        self.damage += other.damage;
    }
}

File: cpp-version/v2/stat.cpp
#pragma once

#include "stat.h"
#include <cassert>

StatDiff::operator bool() const {
	return Flag != 0
		|| HP != 0
		|| ATK != 0
		|| DEF != 0
		|| GR != 0
		|| YK != 0
		|| GK != 0
		|| BK != 0;
}

StatDiff::operator Stat() const {
	return {
		Flag : Flag,
		HP : HP,
		ATK : ATK,
		DEF : DEF,
		GR : GR,
		YK : YK,
		GK : GK,
		BK : BK
	};
}

bool operator >= (const StatDiff& d0, const StatDiff& d1) {
	return (d0.Flag & d1.Flag) == d1.Flag
		&& d0.HP >= d1.HP
		&& d0.ATK >= d1.ATK
		&& d0.DEF >= d1.DEF
		&& d0.GR >= d1.GR
		&& d0.YK >= d1.YK
		&& d0.GK >= d1.GK
		&& d0.BK >= d1.BK;
}

StatDiff& operator += (StatDiff& d0, const StatDiff& d1) {
	d0.Flag |= d1.Flag;
	d0.HP += d1.HP;
	d0.ATK += d1.ATK;
	d0.DEF += d1.DEF;
	d0.GR += d1.GR;
	d0.YK += d1.YK;
	d0.GK += d1.GK;
	d0.BK += d1.BK;
	return d0;
}

StatDiff operator + (const StatDiff& d0, const StatDiff& d1) {
	return StatDiff{
		Flag : d0.Flag | d1.Flag,
		HP : d0.HP + d1.HP,
		ATK : d0.ATK + d1.ATK,
		DEF : d0.DEF + d1.DEF,
		GR : d0.GR + d1.GR,
		YK : d0.YK + d1.YK,
		GK : d0.GK + d1.GK,
		BK : d0.BK + d1.BK
	};
}

StatDiff operator - (const StatDiff& d) {
	assert(!d.Flag);
	return StatDiff{
		Flag : d.Flag,
		HP : - d.HP,
		ATK : - d.ATK,
		DEF : - d.DEF,
		GR : - d.GR,
		YK : - d.YK,
		GK : - d.GK,
		BK : - d.BK
	};
}

// --------------------------------------------------------------------------------

bool Stat::valid() {
	return HP >= 0
		&& ATK >= 0
		&& DEF >= 0
		&& GR >= 0
		&& YK >= 0
		&& GK >= 0
		&& BK >= 0;
}

int Stat::score() const {
	#ifdef TOTS
		return (HP + 1) * 25 + (ATK * 5 + YK * 10 + GK * 20 + BK * 30) * 1000 + DEF * 1000 * 10 / 3;
	#else
		return (HP + 1) * 25 + (ATK * 5 + DEF * 3 + YK * 10 + GK * 20 + BK * 30) * 1000;
	#endif
}

Stat& Stat::operator += (const StatDiff& d) {
	Flag |= d.Flag;
	HP += d.HP;
	ATK += d.ATK;
	DEF += d.DEF;
	GR += d.GR;
	YK += d.YK;
	GK += d.GK;
	BK += d.BK;
	return *this;
}

Stat& Stat::join(const Stat& s) {
	Flag |= s.Flag;
	if (HP < s.HP) { HP = s.HP; };
	if (ATK < s.ATK) { ATK = s.ATK; };
	if (DEF < s.DEF) { DEF = s.DEF; };
	if (GR < s.GR) { GR = s.GR; };
	if (YK < s.YK) { YK = s.YK; };
	if (GK < s.GK) { GK = s.GK; };
	if (BK < s.BK) { BK = s.BK; };
	return *this;
}

Stat Stat::operator - (const StatDiff& d) const {
	return Stat{
		Flag : Flag & ~d.Flag,
		HP : HP - d.HP,
		ATK : ATK - d.ATK,
		DEF : DEF - d.DEF,
		GR : GR - d.GR,
		YK : YK - d.YK,
		GK : GK - d.GK,
		BK : BK - d.BK
	};
}

std::ostream& printScore(std::ostream& os, const Stat& stat) {
	int score = stat.score();
	os << "Score: " << score / 1000 << ".";
	if (score % 1000 < 100) {
		os << "0";
	}
	if (score % 1000 < 10) {
		os << "0";
	}
	return os << score % 1000;
}

std::ostream& printPlayerFlag(std::ostream& os, const FLAG Flag) {
	os << "Flag:";
	if (Flag & DDoubleGR) {
		os << " DoubleGR";
	}
	if (Flag & DDoubleATKAgainstGoblin) {
		os << " DoubleATKAgainstGoblin";
	}
	if (Flag & DDoubleATKAgainstWyrm) {
		os << " DoubleATKAgainstWyrm";
	}
	return os;
}

std::ostream& operator << (std::ostream& os, const Stat& stat) {
	printScore(os, stat) << "\n{ ";
	return printPlayerFlag(os, stat.Flag)
		<< ", HP: " << stat.HP + 1
		<< ", ATK: " << stat.ATK
		<< ", DEF: " << stat.DEF
		<< ", GR: " << stat.GR
		<< ", YK: " << stat.YK
		<< ", GK: " << stat.GK
		<< ", BK: " << stat.BK
		<< "}\n";
}

bool operator >= (const Stat& s0, const Stat& s1) {
	return (s0.Flag & s1.Flag) == s1.Flag
		&& s0.HP >= s1.HP
		&& s0.ATK >= s1.ATK
		&& s0.DEF >= s1.DEF
		&& s0.GR >= s1.GR
		&& s0.YK >= s1.YK
		&& s0.GK >= s1.GK
		&& s0.BK >= s1.BK;
}

// --------------------------------------------------------------------------------

EssStat::EssStat(Stat s) {
	Flag = s.Flag;
	ATK = s.ATK;
	DEF = s.DEF;
}

bool operator == (const EssStat s0, const EssStat s1) {
	return s0.Flag == s1.Flag && s0.ATK == s1.ATK && s0.DEF == s1.DEF;
}

EssStat::operator Stat() const {
	return {
		Flag : Flag,
		ATK : ATK,
		DEF : DEF
	};
}

// --------------------------------------------------------------------------------

ProbeStat& operator += (ProbeStat& s0, const ProbeStat& s1) {
	s0.req.join(s1.req - s0.diff);
	s0.diff += s1.diff;
	s0.loss += s1.loss;
	return s0;
}
